"gradlew buildBwcVersion" fails with misleading error if JAVA11_HOME is not set

**Elasticsearch version** (`bin/elasticsearch --version`): build issue, seen with `master` branch

**Plugins installed**: []

**JVM version** (`java -version`): 

```
openjdk version "11.0.1" 2018-10-16
OpenJDK Runtime Environment (build 11.0.1+13-Ubuntu-2ubuntu1)
OpenJDK 64-Bit Server VM (build 11.0.1+13-Ubuntu-2ubuntu1, mixed mode, sharing)
```

**OS version** (`uname -a` if on a Unix-like system):

```
Linux <name> 4.18.0-13-generic #14-Ubuntu SMP Wed Dec 5 09:04:24 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Description of the problem including expected versus actual behavior**:

When I run `buildBwcVersion`, it fails with `ERROR: JAVA_HOME is set to an invalid directory: null`, but `JAVA_HOME` is set. It's actually `JAVA11_HOME` that's missing.

**Steps to reproduce**:

1. Confirm that `JAVA_HOME` is set, but `JAVA11_HOME` is not (`env | grep JAVA`)
2. Check out the latest source and confirm that Java 11 is set in [java-versions.properties](https://github.com/elastic/elasticsearch/blob/6dcb3af4c8300d2d5f5fc9848c2a039956edda5d/.ci/java-versions.properties) (`grep 11 .ci/java-versions.properties`)
3. Run `./gradlew buildBwcVersion`. The build should fail with `ERROR: JAVA_HOME is set to an invalid directory: null`.

**Provide logs (if relevant)**:

```
> Task :distribution:bwc:bugfix:buildBwcVersion FAILED

 [6.5.5] ERROR: JAVA_HOME is set to an invalid directory: null
 [6.5.5] 
 [6.5.5] Please set the JAVA_HOME variable in your environment to match the
 [6.5.5] location of your Java installation.
 [6.5.5] 

FAILURE: Build failed with an exception.

* What went wrong:

Execution failed for task ':distribution:bwc:bugfix:buildBwcVersion'.
> Process 'command '/home/heather/code/git/elasticsearch/distribution/bwc/bugfix/build/bwc/checkout-6.5/gradlew'' finished with non-zero exit value 1
```

--- a/buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy
+++ b/buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy
@@ -395,7 +395,6 @@ class BuildPlugin implements Plugin<Project> {
             rootProject.rootProject.ext.requiredJavaVersions = [:].withDefault{key -> return []}
             rootProject.gradle.taskGraph.whenReady { TaskExecutionGraph taskGraph ->
                 List<String> messages = []
-                println("Checking for java versions: " + rootProject.requiredJavaVersions.keySet())
                 for (entry in rootProject.requiredJavaVersions) {
                     if (rootProject.javaVersions.get(entry.key) != null) {
                         continue

--- a/buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy
+++ b/buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy
@@ -395,6 +395,7 @@ class BuildPlugin implements Plugin<Project> {
             rootProject.rootProject.ext.requiredJavaVersions = [:].withDefault{key -> return []}
             rootProject.gradle.taskGraph.whenReady { TaskExecutionGraph taskGraph ->
                 List<String> messages = []
+                println("Checking for java versions: " + rootProject.requiredJavaVersions.keySet())
                 for (entry in rootProject.requiredJavaVersions) {
                     if (rootProject.javaVersions.get(entry.key) != null) {
                         continue
@@ -407,9 +408,16 @@ class BuildPlugin implements Plugin<Project> {
                 if (messages.isEmpty() == false) {
                     throw new GradleException(messages.join('\n'))
                 }
+                rootProject.rootProject.ext.requiredJavaVersions = null // reset to null to indicate the pre-execution checks have executed
             }
+        } else if (rootProject.rootProject.requiredJavaVersions == null) {
+            // check directly if the version is present since we are already executing
+            if (rootProject.javaVersions.get(version) == null) {
+                throw new GradleException("JAVA${version}_HOME required to run task:\n${task}")
+            }
+        } else {
+            rootProject.requiredJavaVersions.get(version).add(task)
         }
-        rootProject.requiredJavaVersions.get(version).add(task)
     }
 
     /** A convenience method for getting java home for a version of java and requiring that version for the given task to execute */

